var/spack/repos/builtin/packages/py-scikit-odes/package.py

```diff
@@ -0,0 +1,33 @@
+# Copyright 2013-2023 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+from spack.package import *
+
+
+class PyScikitOdes(PythonPackage):
+    """Odes is a scikit toolkit for scipy to add extra ode solvers.
+    Specifically it interfaces the Sundials solvers cvode, cvodes, ida and
+    idas.  It this way it provides extra modern ode and dae solvers you can
+    use, extending the capabilities offered in scipy.integrade.ode."""
+
+    homepage = "https://github.com/bmcage/odes"
+
+    pypi = "scikits.odes/scikits.odes-2.7.0.tar.gz"
+    github = "https://github.com/bmcage/odes.git"
+
+    maintainers("omsai")
+
+    version("2.7.0", sha256="a71e19e1485893754ae8c050668232fcc694f17b83602e75fbebf7bf9f975e1e")
+
+    depends_on("python@3.7:", type=("build", "run"))
+
+    depends_on("py-setuptools@:64.0.0", type="build")
+
+    depends_on("py-numpy", type=("build", "run"))
```

Comment:
The requirements were made to match the [pyproject.toml](https://github.com/bmcage/odes/blob/v2.7.0/pyproject.toml)

Reply:
I'm doing a build on an fresh spack clone with the oldest sundials@5.1 to determine if the pyproject.toml is improper.

Reply:
Thank you - I added the py-scipy runtime dependency and removed the py-numpy dependency and confirmed the unit tests pass:

```console
$ spack unload --all
$ spack env create skodes
$ spack env activate skodes
$ spack add py-scikit-odes ^py-numpy@:1.23  # to use the now removed numpy.testing.run_module_suite
$ spack add py-pytest
$ spack install
$ spack load python
$ export SPACK_PYTHON=`which python`  # necessary for py-pytest on Debian-ish systems according to Harmen Stoppels
$ source share/spack/setup-env.sh
$ spack load
$ spack cd -i py-scikit-odes
$ cd lib/python3.10/site-packages/scikits/odes/tests/
$ pytest -W ignore --durations-min=1
============================= test session starts ==============================
platform linux -- Python 3.10.12, pytest-7.3.2, pluggy-1.0.0
rootdir: /home/omsai/src/spack-py-scikit-odes
configfile: pytest.ini
collected 126 items                                                            

test_dae.py .......                                                      [  5%]
test_dop.py .                                                            [  6%]
test_get_info.py .....                                                   [ 10%]
test_odeint.py .....                                                     [ 14%]
test_on_funcs.py ............                                            [ 23%]
test_on_funcs_ida.py ............                                        [ 33%]
test_user_return_vals_cvode.py ......................................... [ 65%]
.......                                                                  [ 71%]
test_user_return_vals_ida.py ....................................        [100%]

============================= slowest 30 durations =============================

(30 durations < 1s hidden.  Use -vv to show these durations.)
============================= 126 passed in 1.77s ==============================
```

Note to run the unit tests I have to use `py-scikit-odes ^py-numpy@:1.23` otherwise I get `ImportError: cannot import name 'run_module_suite' from 'numpy.testing'`.

Reply:
Sorry, I overlooked the `py-numpy` build dependency. According to the recipe in `pyproject.toml` and `setup.py` it should be added as a build dependency only. But after your findings with the testing I did a quick search in the source code and found `import numpy` statements also elsewhere (e.g. in [scikits/odes/ode.py#L29](https://github.com/bmcage/odes/blob/v2.7.0/scikits/odes/ode.py#L29) so it seems that this is an error in the recipe. Let's just add it as build and run dependency (as you had before) and add a comment why that is.

Regarding the version: this does only affect the tests and there is already a [merged pull request](https://github.com/bmcage/odes/pull/153) for it. By adding this PR as patch the problem is solved for me.

BTW you can add the tests also into the spack recipe directly, then every time you install the package with `spack install --test=root py-scikit-odes` the tests are executed as well. I did it just now like this:

```
    # Remove numpy test runner imports to be compatible with `py-numpy@1.25:`
    patch("https://github.com/bmcage/odes/pull/153.patch",
          sha256="8d05d7bcc3582b7c482a4393bf5a8c0460a58eb62d1e3c86339c95a0d4ce30ac")

    depends_on("py-pytest", type="test")

    @run_after("install")
    @on_package_attributes(run_tests=True)
    def install_test(self):
        with working_dir("spack-test", create=True):
            pytest = which("pytest")
            pytest(join_path(self.prefix, python_purelib, "scikits", "odes", "tests"))
```

Reply:
Thank you; I've fixed the numpy and scipy dependencies, added the PR patch you found, and added the test function.